README.rst
sentinel-value
==============

|pypi badge| |build badge| |docs badge|


``sentinel-value`` is a Python package, that helps to create `Sentinel Values`_ -
special singleton objects, akin to ``None``, ``NotImplemented`` and  ``Ellipsis``.

It implements the ``sentinel()`` function (described by `PEP 661`_),
and for advanced cases it also provides the ``SentinelValue()`` class (not a part of `PEP 661`_).

.. _`Sentinel Values`: https://en.wikipedia.org/wiki/Sentinel_value
.. _`PEP 661`: https://www.python.org/dev/peps/pep-0661


Usage example:

.. code:: python

  from sentinel_value import sentinel

  MISSING = sentinel("MISSING")

  def get_something(default=MISSING):
      ...
      if default is not MISSING:
          return default
      ...


Or, the same thing, but using the ``SentinelValue`` class
(slightly more verbose, but allows to have nice type annotations):

.. code:: python

  from typing import Union
  from sentinel_value import SentinelValue

  class Missing(SentinelValue):
      pass

  MISSING = Missing(__name__, "MISSING")

  def get_something(default: Union[str, Missing] = MISSING):
      ...
      if default is not MISSING:
          return default
      ...


Links
-----

- Read the Docs: https://sentinel-value.readthedocs.io
- GitHub repository: https://github.com/vdmit11/sentinel-value
- Python package: https://pypi.org/project/sentinel-value/


.. |pypi badge| image:: https://img.shields.io/pypi/v/sentinel-value.svg
  :target: https://pypi.org/project/sentinel-value/
  :alt: Python package version

.. |build badge| image:: https://github.com/vdmit11/sentinel-value/actions/workflows/build.yml/badge.svg
  :target: https://github.com/vdmit11/sentinel-value/actions/workflows/build.yml
  :alt: Tests Status

.. |docs badge| image:: https://readthedocs.org/projects/sentinel-value/badge/?version=latest
  :target: https://sentinel-value.readthedocs.io/en/latest/?badge=latest
  :alt: Documentation Status



CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v1.0.0 (2021-11-29)
### Feature
* Reverse parameters of SentinelValue() constructor ([`0a3b77f`](https://github.com/vdmit11/sentinel-value/commit/0a3b77f624f8da0909130381b0a6c5aa9f92e12e))

### Breaking
* SentinelValue() parameter order is different now: ([`0a3b77f`](https://github.com/vdmit11/sentinel-value/commit/0a3b77f624f8da0909130381b0a6c5aa9f92e12e))

## v0.3.2 (2021-11-03)
### Fix
* Py.typed is installed to site-packages directory ([`ffa8256`](https://github.com/vdmit11/sentinel-value/commit/ffa8256c02cf691241dee979a3e38be387206b6d))

## v0.3.1 (2021-11-03)
### Fix
* Missing py.typed marker file (see PEP 561) ([`c5bcd5a`](https://github.com/vdmit11/sentinel-value/commit/c5bcd5a5ce1154a2312701469b16fa4831b73494))

## v0.3.0 (2021-11-03)
### Feature
* Initial stable release, remove under-development warning ([`92f1f6f`](https://github.com/vdmit11/sentinel-value/commit/92f1f6f5ad81bd26af4c29c861608f8ffcccc814))


run_tests.sh
#!/usr/bin/env sh

echo '----------------'

set -e
set -x

pytest --cov=sentinel_value --cov-fail-under=100 $@


test_sentinel_value.py
from unittest import mock

import pytest

from sentinel_value import SentinelValue, sentinel, sentinel_value_instances


@pytest.fixture(autouse=True)
def cleanup_sentinel_value_instances_after_each_test():
    yield
    sentinel_value_instances.clear()


def test__SentinelValue__str_and_repr():
    MISSING = SentinelValue("MISSING", "some.module2")
    assert str(MISSING) == "<MISSING>"
    assert repr(MISSING) == "<MISSING>"


def test__SentinelValue__is_falsy():
    MISSING = SentinelValue("MISSING", "some.module3")
    assert not bool(MISSING)


def test__SentinelValue_creates_new_instance_only_once_per_name():
    MISSING = SentinelValue("MISSING", "some.module")

    MISSING_duplicate = SentinelValue("MISSING", "some.module")
    assert MISSING is MISSING_duplicate

    MISSING2 = SentinelValue("MISSING2", "some.module")
    assert MISSING is not MISSING2

    MISSING_in_other_module = SentinelValue("MISSING", "some.module2")
    assert MISSING is not MISSING_in_other_module


def test__SentinelValue_can_change_class_on_the_fly():
    MISSING = SentinelValue("MISSING", __name__)
    assert MISSING.__class__ is SentinelValue

    class Missing(SentinelValue):
        pass

    MISSING_redefined = Missing("MISSING", __name__)

    assert MISSING_redefined is MISSING
    assert MISSING_redefined.__class__ is Missing


def test__sentinel__custom_repr():
    MISSING1 = sentinel("MISSING1")
    assert repr(MISSING1) == "<MISSING1>"

    MISSING2 = sentinel("MISSING2", repr="my.module.MISSING2")
    assert repr(MISSING2) == "my.module.MISSING2"


def test__sentinel__can_be_called_multiple_times():
    MISSING = sentinel("MISSING")
    MISSING_duplicate = sentinel("MISSING")

    assert MISSING_duplicate is MISSING


@mock.patch("inspect.currentframe")
def test__sentinel__throws_error_if_call_stack_not_available(currentframe_mock):
    currentframe_mock.return_value = None

    with pytest.raises(AssertionError):
        sentinel("MISSING")


def test__sentinel__creates_subclass_of_SentinelValue():
    MISSING = sentinel("MISSING")
    Missing = MISSING.__class__

    assert Missing is not SentinelValue
    assert issubclass(Missing, SentinelValue)

    MISSING2 = sentinel("MISSING")
    Missing2 = MISSING2.__class__

    assert Missing2 is Missing


tests/test_sentinel_value.py
import pickle
from copy import copy, deepcopy
from unittest import mock

import pytest

from sentinel_value import SentinelValue, sentinel, sentinel_value_instances


@pytest.fixture(autouse=True)
def cleanup_sentinel_value_instances_after_each_test():
    yield
    sentinel_value_instances.clear()


def test__SentinelValue__str_and_repr():
    MISSING = SentinelValue("some.module2", "MISSING")
    assert str(MISSING) == "<MISSING>"
    assert repr(MISSING) == "<MISSING>"


def test__SentinelValue__is_falsy():
    MISSING = SentinelValue("some.module3", "MISSING")
    assert not bool(MISSING)


def test__SentinelValue_creates_new_instance_only_once_per_name():
    MISSING = SentinelValue("some.module", "MISSING")

    MISSING_duplicate = SentinelValue("some.module", "MISSING")
    assert MISSING is MISSING_duplicate

    MISSING2 = SentinelValue("some.module", "MISSING2")
    assert MISSING is not MISSING2

    MISSING_in_other_module = SentinelValue("some.module2", "MISSING")
    assert MISSING is not MISSING_in_other_module


def test__SentinelValue_can_change_class_on_the_fly():
    MISSING = SentinelValue(__name__, "MISSING")
    assert MISSING.__class__ is SentinelValue

    class Missing(SentinelValue):
        pass

    MISSING_redefined = Missing(__name__, "MISSING")

    assert MISSING_redefined is MISSING
    assert MISSING_redefined.__class__ is Missing


def test__SentinelValue_is_pickleable():
    MISSING = SentinelValue(__name__, "MISSING")
    MISSING_unpickled = pickle.loads(pickle.dumps(MISSING))
    assert MISSING_unpickled is MISSING


class NoValue(SentinelValue):
    pass


def test__SentinelValueSubclass_is_pickleable():
    NO_VALUE = NoValue(__name__, "NO_VALUE")
    NO_VALUE_unpickled = pickle.loads(pickle.dumps(NO_VALUE))
    assert NO_VALUE_unpickled is NO_VALUE


def test__SentinelValue__remains_singleton__when_copied():
    class Missing(SentinelValue):
        pass

    DELETED = Missing(__name__, "DELETED")
    NEVER_SET = Missing(__name__, "NEVER_SET")

    DELETED_copy = copy(DELETED)
    assert DELETED_copy is DELETED

    dct = {"value1": DELETED, "value2": NEVER_SET}
    dct_copy = deepcopy(dct)

    assert dct_copy["value1"] is DELETED
    assert dct_copy["value2"] is NEVER_SET


def test__instance_produced_by_sentinel_function__is_picklable():
    NOT_GIVEN = sentinel("NOT_GIVEN")
    NOT_GIVEN_unpickled = pickle.loads(pickle.dumps(NOT_GIVEN))
    assert NOT_GIVEN_unpickled is NOT_GIVEN


def test__sentinel__custom_repr():
    MISSING1 = sentinel("MISSING1")
    assert repr(MISSING1) == "<MISSING1>"

    MISSING2 = sentinel("MISSING2", repr="my.module.MISSING2")
    assert repr(MISSING2) == "my.module.MISSING2"


def test__sentinel__can_be_called_multiple_times():
    MISSING = sentinel("MISSING")
    MISSING_duplicate = sentinel("MISSING")

    assert MISSING_duplicate is MISSING


@mock.patch("inspect.currentframe")
def test__sentinel__throws_error_if_call_stack_not_available(currentframe_mock):
    currentframe_mock.return_value = None

    with pytest.raises(AssertionError):
        sentinel("MISSING")


def test__sentinel__creates_subclass_of_SentinelValue():
    MISSING = sentinel("MISSING")
    Missing = MISSING.__class__

    assert Missing is not SentinelValue
    assert issubclass(Missing, SentinelValue)

    MISSING2 = sentinel("MISSING")
    Missing2 = MISSING2.__class__

    assert Missing2 is Missing


sentinel_value/sentinel_value.py
import inspect
import sys
import threading
from typing import Dict, Optional, Type


class SentinelValue:
    """Class for special unique placeholder objects, akin to :data:`None` and :data:`Ellipsis`.

    Useful for distinguishing "value is not set" and "value is set to None" cases
    as shown in this example::

        >>> NOT_SET = SentinelValue(__name__, "NOT_SET")

        >>> value = getattr(object, "some_attribute", NOT_SET)
        >>> if value is NOT_SET:
        ...    print('attribute is not set')
        attribute is not set

    If you need a separate type (for use with :mod:`typing` or :func:`functools.singledispatch`),
    then you can create a subclass::

        >>> from typing import Union

        >>> class Missing(SentinelValue):
        ...     pass

        >>> MISSING = Missing(__name__, "MISSING")

        # Here is how the Missing class can be used for type hinting.
        >>> value: Union[str, None, Missing] = getattr(object, "some_attribute", MISSING)
        >>> if value is MISSING:
        ...    print("value is missing")
        value is missing
    """

    def __init__(self, module_name: str, instance_name: str) -> None:
        """Initialize :class:`SentinelValue` object.

        :param module_name: name of Python module that hosts the sentinel value.
                            In the majority of cases you should pass ``__name__`` here.
        :param instance_name: name of Python variable that points to the sentinel value.
        """
        assert not ((module_name == self.__module__) and (instance_name == self.__class__.__name__))

        self.instance_name = instance_name
        self.module_name = module_name
        self.qualified_name = self._compose_qualified_name(instance_name, module_name)

        super().__init__()

    def __new__(cls, module_name, instance_name):
        """Create 1 instance of SentinelValue per name.

        Usually, when you call a class, you expect to get a new instance on each call.

        This is not true for the :class:`SentinelValue` class, that overrides :meth:`object.__new__`
        method to alter the way new instances are created.

        The overridden :meth:`SentinelValue.__new__` method constructs 1 unique instance per name,
        then saves it in the global registry called :data:`sentinel_value_instances`,
        and then next time it returns an already existing object.

        That is, if you call :class:`SentinelValue` multiple times with the same arguments,
        you get the *exactly same* instance, check this out::

          >>> MISSING1 = SentinelValue(__name__, "MISSING")
          >>> MISSING2 = SentinelValue(__name__, "MISSING")

          >>> MISSING1 is MISSING2
          True

        That guarantees that for each name, there is only 1 unique sentinel object.

        This is needed for 2 things:

        1. :mod:`pickle` - when a sentinel value is pickled/un-pickled,
           you get the same instance (not a copy).

        2. Hot code reloading

           Many IDEs can re-load individual modules on the fly, without re-starting
           the whole Python process. So this uniqueness trick allows to avoid duplicate instances
           that would  appear if the module (containing sentinel value definition) is re-loaded.

           This is also nice when working with a Python shell.
           You can re-send ``SentinelValue()`` calls over and over again to the shell,
           and that wouldn't break your code (that already references the instance).

        """
        qualified_name = cls._compose_qualified_name(instance_name, module_name)

        # The create-if-not-exists kind operation has to be protected with a lock.
        #
        # Otherwise, two (or more) concurrent threads will both see that a sentinel value
        # doesn't exist yet, both will create a new instance, and thus you get a duplicate.
        #
        # Well, of course, there is GIL (Global Interpreted Lock), that, in theory,
        # should protect us as long as we're not doing any I/O, but, there are no guarantees
        # about the GIL, and who knows what would # happen in the future.
        #
        # So let's just not rely on GIL, and just use 1 custom lock here,
        # that guarantees that a new object is really created only once.
        with sentinel_create_lock:
            # Check if the instance already exists in the global registry.
            existing_instance = sentinel_value_instances.get(qualified_name)
            if existing_instance is not None:
                # Change class on the fly. This is needed hot live code reloading features.
                # That is, if a subclass of SentinelValue was re-defined,
                # then we want to switch to the new class (but still use the old instance).
                # A bit hacky, but works.
                if existing_instance.__class__ is not cls:
                    existing_instance.__class__ = cls
                return existing_instance

            # Ok, no existing instance.
            # Then create a new one, and put it to the global registry.
            new_instance = super().__new__(cls)
            sentinel_value_instances[qualified_name] = new_instance
            return new_instance

    def __getnewargs__(self):
        # Get arguments for the __new__ method.
        # This is needed for pickle serialization.
        # In combination with magic in the overriden __new__() method above,
        # that allows to avoid constructing duplicates when un-pickling the object.
        return (self.module_name, self.instance_name)

    @staticmethod
    def _compose_qualified_name(instance_name: str, module_name: str) -> str:
        return module_name + "." + instance_name

    def __repr__(self):
        """Provide :func:`repr` for :class:`SentinelValue`.

        By default, looks like this::

            <MISSING>

        You're free to override it in a subclass if you want to customize it.
        """
        return "<" + self.instance_name + ">"

    @staticmethod
    def __bool__():
        """Return False when :class:`SentinelValue` is treated as :func:`bool`.

        Sentinel values are always falsy.

        This is done because most sentinel objects are "no value" kind of objects
        (they're like ``None``, but just not the ``None`` object).

        So it is often handy to do ``if not value`` to check if there is no value
        (like if an attribute is set to ``None``, or not set at all)::

           >>> NOT_SET = SentinelValue(__name__, "NOT_SET")

           >>> value = getattr(object, "foobar", NOT_SET)

           # Is the value None, or empty, or not set at all?
           >>> if not value:
           ...    print("no value")
           no value

        If this doesn't fit your case, you can override this method in a subclass.
        """
        return False


sentinel_value_instances: Dict[str, SentinelValue] = {}
"""Dictionary that contains all instances of SentinelValue (and its subclasses).

This dictionary looks like this::

  {
      "package1.module1.MISSING": SentinelValue("package1.module1", "MISSING"),
      "package2.module2.MISSING": SentinelValue("package2.module2", "MISSING"),
      "package2.module2.ABSENT": SentinelValue("package2.module2", "ABSENT"),
  }

When a :class:`SentinelValue` object is instanciated, it registers itself in this dictionary
(and throws an error if already registered). This is needed to ensure that, for each name,
there exists only 1 unique :class:`SentinelValue` object.
"""


sentinel_create_lock = threading.Lock()
"""A lock that prevents race conditions when creating new :class:`SentinelValue` objects.

Problem: when you start multiple threads, they may try to create sentinel objects concurrently.
If you're lucky enough, you get duplicate :class:`SentinelValue` instances,
which is highly undesirable.

This :data:`sentinel_create_lock` helps to protect against such race conditions.

The lock is acquired whenever a new :class:`SentienlValue` object is created.
So, when multiple threads try to create sentinel objects, then they're executed in sequence,
and the 1st thread really creates a new instance, and other threads will get the already
existing instance.
"""


def sentinel(
    instance_name: str,
    repr: Optional[str] = None,
) -> SentinelValue:
    """Create an unique sentinel object.

    Implementation of PEP 661
    https://www.python.org/dev/peps/pep-0661/

        >>> MISSING = sentinel("MISSING")

        >>> value = getattr(object, "value", MISSING)

        >>> if value is MISSING:
        ...     print("value is not set")
        value is not set

    :param instance_name: Name of Python variable that points to the sentinel object.
                          Needed for serialization (like :mod:`pickle`) and also nice :func:`repr`.

    :param repr: Any custom string that will be returned by func:`repr`.
                 By default, composed as ``{module_name}.{instance_name}``.
    """
    # pylint: disable=redefined-builtin

    module_name = _get_caller_module_name()
    assert module_name

    SentinelValueSubclass = _create_sentinel_value_subclass(module_name, instance_name)

    if repr:
        SentinelValueSubclass.__repr__ = lambda self: repr  # type: ignore

    return SentinelValueSubclass(module_name, instance_name)


def _get_caller_module_name() -> Optional[str]:
    # Walk over the call stack, and stop as soon as we leave this (sentinel_value) module.
    frame = inspect.currentframe()
    while frame:
        module_name: str = frame.f_globals["__name__"]

        if module_name != __name__:
            return module_name

        frame = frame.f_back

    # Normally the code should never reach this point.
    # It may be only the case when stack inspectio is not available
    # (on some alternative Python implementation, like maybe Jython)
    return None


def _create_sentinel_value_subclass(module_name: str, instance_name: str) -> Type[SentinelValue]:
    module = sys.modules[module_name]

    # Genarate class name from variable name.
    # E.g.: MISSING -> _sentinel_MISSING
    class_name = "_sentinel_" + instance_name.replace(".", "_")

    # Class should be created only once, so first check if it was already created.
    if hasattr(module, class_name):
        existing_class: Type[SentinelValue] = getattr(module, class_name)
        assert issubclass(existing_class, SentinelValue)
        return existing_class

    # Create a new subclass of SentinelValue.
    SentinelValueSubclass = type(class_name, (SentinelValue,), {})

    # Bind class with the module.
    # That modifies module's globals, so the class really becomes a member of the module,
    # indistinguishable from classes that you define in the code.
    setattr(module, class_name, SentinelValueSubclass)
    SentinelValueSubclass.__module__ = module_name

    return SentinelValueSubclass
